README.md
# Real Time Logs 

Real Time Logs is a node.js app designed for reading logs in real time via ssh and serving them in real time to clients via web sockets.

## Installation and Setup
1. Get clone and cd into directory of your choice
2. npm install
3. Edit or create '/etc/rtlogs/config.js' and add servers and groups.
4. node app.js 

Don't forget to make sure the account you run the app as has correct ssh keys setup. 

sample config.js
```javascript
exports.servers = [
	{
	group: 'static', 
	server: 'static-001', 
	log: '/var/log/Software/static.log'
	},
	{
	group: 'static',
	server: 'static-002',
	log: '/var/log/Software/static.log'
	},
	{
	group: 'api',
	server: 'api-001',
	log: '/var/log/Software/api.log'
	},
	{
	group: 'api',
	server: 'api-002',
	log: '/var/log/Software/api.log'
	},
	{
	group: 'auth',
	server: 'auth-001',
	log: '/var/log/Software/auth.log'
	},
];
```
### Log syntax
Right now there is a pretty definitive log format of 'date space server space ip space message' but this will hopefully change in the future.

### Screenshot
![Screen Shot](http://i.imgur.com/9lh8c.png)

### Special Credit
Graphs by [cubism](https://github.com/square/cubism/)

History.md
0.1.6 / 2012-08-29
==================

  * changed; console.error STDERR for filewatchers
  * changed; automatically retry connections on failure
  * fixed; don't leak fileWatcher to global in streamLog

public/js/core.js
"use strict";

jQuery(function($) {
	window.graphEnabled = false;

	if (localStorage.hideip) {
		$("#hide-ip").val(localStorage.getItem("hideip"))
		console.log("read " + localStorage.getItem("hideip"))
	}

	var lineNum = 0;
	window.linenum = lineNum;
	var socket = io.connect('',{reconnect: true});
	var uniqueServers = [];

	socket.on('log', parseLine);

	function urlize(text) {
		var exp = /(\b(https?|ftp|file):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/ig;
		return ("" + text).replace(exp, '<a href="$1" rel="nofollow" target="_blank">$1</a>');
	};

	function isIP(text) {
		return ("" + text).match(/[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/);
	};

	function addCommas(nStr) {
		var x = ("" + nStr).split('.');
		var x1 = x[0],
				x2 = !!x[1] ? ('.' + x[1]) : '';

		x1 = x1.replace(/(\d+)(\d{3})/g, '$1,$2');
		return x1 + x2;
	}

	//ParLine and LogLine should be more readable and well defined.
	function parseLine(data) {
		currentLineCount = data.lineCount

		$('.numMessages').html(addCommas(currentLineCount) + '<span class="subtext"> messages</span>');

		var temp = ("" + data.text).split(' '),
				tmpIP = " ";

		temp.forEach(function(maybeIP) {
			if(isIP(maybeIP)) {
				tmpIP = maybeIP;
				return false;
			}
		});

		logLine(data.group, data.lineCount, data.server, temp.slice(0, 3).join(' '), tmpIP, 'type', data.text)
	}

	function logLine(group, lineCount, server, date, ip, type, message){
		var ipArray = $('#hide-ip').val().split(",");

		if ($.inArray(ip, ipArray) == -1) {
			var msg = urlize(message.replace(date, '').replace(ip, '').replace(/-\s(?!:-)/g, ''));
			var line = '<tr class="lineItem ' + group + '">' +
			'<td class="lineNum">' + addCommas(lineCount) + '</td>' +
			'<td>' + date + '</td>' +
			'<td>' + server + '</td>' +
			'<td><a target=_blank href="http://ip-address-lookup-v4.com/ip/' + ip + '">' + ip + '</a></td>' +
			'<td class="msg">' + msg + '</td>' +
			'</tr>';
			$('#datagrid').prepend(line);
			if (lineNum > 1000) {
		  	$('.lineItem').last().remove();
		  }

		  lineNum++;
		  window.linenum = lineNum;
		}
	}

	//Date and Time stuff here
	function two(x) {return ((x>9)?"":"0")+x}
	function three(x) {return ((x>99)?"":"0")+((x>9)?"":"0")+x}

	function time(ms) {
		var sec = Math.floor(ms/1000)
		ms = ms % 1000
		//t = three(ms)
		var t = ""

		var min = Math.floor(sec/60)
		sec = sec % 60
		t = two(sec) + ":" + t

		var hr = Math.floor(min/60)
		min = min % 60
		t = two(min) + ":" + t

		var day = Math.floor(hr/60)
		hr = hr % 60
		t = two(hr) + ":" + t
		t = day + ":" + t

		return t
	}

	var currentLineCount = 0;
	var connectMS = 0;

	setInterval(function() {
		connectMS += 1000
		$('.elapsed').html(time(connectMS) + '<span class="subtext"> servers</span>');
		$('.mps').html(Math.round(currentLineCount / (connectMS / 60000) ) + '<span class="subtext"> MPM</span>');
		window.graphEnabled = true; //Wait one second until all the start up data is done loading to start graphing.
	}, 1000);

	//
	socket.on('system', function (data) {
		if(data && data.time) connectMS = data.time;

		if(data && data.servers) {
			var serverGroups = _.pluck(data.servers, 'group');
			uniqueServers.push.apply(uniqueServers, serverGroups);


			$('.serverCount').html(data.servers.length + '<span class="subtext"> servers</span>');
			$('.elapsed').html(time(connectMS) + '<span class="subtext"> elapsed</span>');
		}

		var serverGroups = _.uniq(uniqueServers),
				groupCSS = "";

		serverGroups.forEach(function (elem){
			if ($('#' + elem).size() === 0) {
				$('.footer').append('<div data-group="' + elem + '" class="serversGroup">' + elem + '</div>');
				groupCSS += '#master.' + elem + ' tr.' + elem + ' { display: table-row; }\n';
			}
		});

		$("body").append('<style>' + groupCSS + '</style>');
	});

	socket.on('buffer', function (result) {
		(result.buffer || []).forEach(parseLine)
	});


	$('.footer').on('click', '.serversGroup', function() {
		$('.footer').find('.active').removeClass('active');
		$(this).addClass('active');

		$("#master").removeClass().addClass($(this).data('group') || 'all');
	});

	$(window).resize(function() {
		var autoMasterHeight = $(window).height() - ($(".topbar").height() + $(".footer").height()),
				autoCanvasWidth = Math.min(($(".topbar").width() - $(".topbarWrapper").width()) - 1, 1000);

		$("#master").css({ height: autoMasterHeight });
		$("canvas").css({ width: autoCanvasWidth, height: 40 });
	});

	$(window).resize();

	$("#save").on("click", function(){
		console.log("saved " + $('#hide-ip').val())
		localStorage.setItem("hideip", $('#hide-ip').val() )
	})
});


public/js/graphing.js
"use strict";

jQuery(function($) {
	var lastLineNum = 0;

	function graphData(name) {
		var value = 0,
				values = [],
				i = 0,
				last;

		return context.metric(function(start, stop, step, callback) {
			start = +start, stop = +stop;
			if (isNaN(last)) last = start;

			while (last < stop) {
				last += step;
				value = window.linenum - lastLineNum;
				lastLineNum = window.linenum;
				if (window.graphEnabled) values.push(value);
			}

			callback(null, values = values.slice((start - stop) / step));
		}, name);
	}

	var topbarSize = $(".topbarWrapper").width()
	var autoWidth = Math.round(window.innerWidth - topbarSize - 200)

	var context = cubism.context()
			.serverDelay(0)
			.clientDelay(0)
			.step(1e3)
			.size(autoWidth);

	var foo = graphData("foo");

	d3.select("#graphBar").call(function(div) {
		div.datum(foo);

		div.append("div")
				.attr("class", "horizon")
				.call(context.horizon()
					.height(40)
					.colors(["#FFFFFF","#FFFFFF","#FFFFFF","#FFFFFF","#00CC33","#BBFF66","#FFBB00","#FF0000"])
					.title(" "));
	});

	// On mousemove, reposition the chart values to match the rule.
	context.on("focus", function(i) {
		d3.selectAll(".value").style("right", i == null ? null : context.size() - i + "px");
	});
});


app.js
var http = require('http'),
		socketio = require('socket.io'),
		fs = require('fs'),
		util = require('util'),
		spawn = require('child_process').spawn,
		connect = require('connect');

var lineCount = 1,
		buffer = [],
		startTime = new Date();

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR   = 60 * MINUTE;
const DAY    = 24 * HOUR;

///// Config file stuff
var config;
!function requireAndValidateConfig() {
	try {
		config = require('/etc/rtlogs/config.js');
	} catch(e) {
		console.error('Could not load configuration file [ %s ]', '/etc/rtlogs/config.js');
		process.exit(0);
	} finally {
		var validConfig = (config && Array.isArray(config.servers) && config.servers.length > 0);
		if(!validConfig) {
			console.error('Configuration file has no servers defined');
			process.exit(0);
		}
	}
}();

config.servers.forEach(function(elem) {
	streamLog(elem.server, elem.group, elem.log)
	console.log(elem.server)
});

///// Server creation stuff
var connect = connect(
	connect.static(__dirname + '/public/')
);

var app = http.createServer(connect);
app.listen(8080);

var io = socketio.listen(app);
io.set("log level", 1); //Set the log level of socket.io

io.sockets.on('connection', function(socket) {
	var packagedTime = new Date();

	packagedTime.setTime(Date.now() - startTime.getTime());

	socket.emit('system', {
		text: 'Connection Accepted',
		server: 'socket.io',
		group: 'socket.io'
	});

	socket.emit('system', {
		servers: config.servers,
		time: packagedTime.getTime()
	})

	socket.emit('buffer', { buffer: buffer })

	socket.on('my other event', function(data) {
		console.log(data);
	});
});

///// Actual Streamer
function setupFileWatcher(f, server, group, file) {
	f.stdout.setEncoding('utf8');

	f.stdout.on('data', function(data) {
		var lines = ("" + data).split('\n');

		lines.forEach(function(line) {
			if(line === "") return;

			var logEntry = { text: line, server: server, group: group, lineCount: lineCount };

			io.sockets.emit('log', logEntry);
			buffer.push(logEntry);

			if (buffer.length > 100) buffer = buffer.slice(-100);

			lineCount++;
		});
	});

	f.stderr.setEncoding('utf8');

	f.stderr.on('data', function(data) {
		console.error('FileWatcher [ %s ] [ %s ] ERROR: %s', server, group, data);
	});

	f.on('exit', function(code) {
		f = null;
		console.log("ERROR: file watcher died " + code);
		streamLog(server, group, file);
	});
};

function streamLog(server, group, file) {
	var fileWatcher = spawn('ssh', [server, 'tail', '-f', '-n', '0', file]);
	setupFileWatcher(fileWatcher, server, group, file);
}